pkg/virt-controller/watch/replicaset.go

```diff
@@ -290,131 +337,224 @@ func (c *VMReplicaSet) scale(rs *virtv1.VirtualMachineReplicaSet, vms []virtv1.V
+		// and after such time has passed, the virt-handler actually deletes it from the store. We receive an update
+		// for modification of the deletion timestamp and expect an rs to create more replicas asap, not wait
+		// until the virt-handler actually deletes the vm. This is different from the Phase of a vm changing, because
+		// an rs never initiates a phase change, and so is never asleep waiting for the same.
+		c.deleteVirtualMachine(curVM)
+		if labelChanged {
+			// we don't need to check the oldVM.DeletionTimestamp because DeletionTimestamp cannot be unset.
+			c.deleteVirtualMachine(oldVM)
+		}
 		return
 	}
 
-	// In case the controller is waiting for a deletion, tell it that we observed one
-	c.expectations.DeletionObserved(rsKey, controller.VirtualMachineKey(vm))
-	c.Queue.Add(rsKey)
-	return
-}
+	curControllerRef := controller.GetControllerOf(curVM)
+	oldControllerRef := controller.GetControllerOf(oldVM)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged && oldControllerRef != nil {
+		// The ControllerRef was changed. Sync the old controller, if any.
+		if rs := c.resolveControllerRef(oldVM.Namespace, oldControllerRef); rs != nil {
+			c.enqueueReplicaSet(rs)
+		}
+	}
 
-// deleteVirtualMachine searchs for a matching VMReplicaSet and wakes it up
-func (c *VMReplicaSet) updateVirtualMachine(old, curr interface{}) {
-	rsKey := c.getMatchingControllerKey(curr.(*virtv1.VirtualMachine))
-	if rsKey == "" {
+	// If it has a ControllerRef, that's all that matters.
+	if curControllerRef != nil {
+		rs := c.resolveControllerRef(curVM.Namespace, curControllerRef)
+		if rs == nil {
+			return
+		}
+		log.Log.V(4).Object(curVM).Infof("VirtualMachine updated")
+		c.enqueueReplicaSet(rs)
+		// TODO: MinReadySeconds in the VM will generate an Available condition to be added in
+		// Update once we support the available conect on the rs
 		return
 	}
 
-	c.Queue.Add(rsKey)
-	return
+	// Otherwise, it's an orphan. If anything changed, sync matching controllers
+	// to see if anyone wants to adopt it now.
+	if labelChanged || controllerRefChanged {
+		rss := c.getMatchingControllers(curVM)
+		if len(rss) == 0 {
+			return
+		}
+		log.Log.V(4).Object(curVM).Infof("Orphan VirtualMachine updated")
+		for _, rs := range rss {
+			c.enqueueReplicaSet(rs)
+		}
+	}
+}
+
+// When a vm is deleted, enqueue the replica set that manages the vm and update its expectations.
+// obj could be an *v1.VirtualMachine, or a DeletionFinalStateUnknown marker item.
+func (c *VMReplicaSet) deleteVirtualMachine(obj interface{}) {
+	vm, ok := obj.(*virtv1.VirtualMachine)
+
+	// When a delete is dropped, the relist will notice a vm in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the vm
+	// changed labels the new ReplicaSet will not be woken up till the periodic resync.
```

Comment:
Is this comment saying there's a possibility a deletion won't be observed somehow? That would lead the RS to block indefinitely on scaling would it not?

Reply:
It would work as expected once the next wake-up for the controller is triggered. Let me make sure, that we wake our  RS with the same interval like k8s does. However, whenever the VMRS or one of its managed VMs changes, it will be woken and it can adapt. It would just miss that one wake.

Reply:
How would the RS waking up resolve this?  The expected deletion would never be observed in this deleteVirtualMachine function, which means the RS wouldn't be able to scale. 

```
needsSync := c.expectations.SatisfiedExpectations(key)
``
^^^ That needsSync bool in the execute() function would forever return false for the VMRC because the deletion was not observed. Right?

Reply:
Good observation. There exists a timeout in the expectations manager. If there are communication errors which will lead to such a scenario, it can blocked the old controller for 5 minutes max.

Reply:
btw. kubernetes has that on every controller (the comment is added to controllers since at least May 2015). I am not sure if that comment is still valid with shared informers which basically watch every object, and owner references.

Reply:
as long as the expectations manager has a timeout, this is fine.